File: Document/beginners/sample_mruby/sample05/app/tracer.rb
class Tracer

  PWM_MAX = 100
  PWM = PWM_MAX / 6
  THRESHOLD = 20

  # コンストラクタ
  #
  # 引数でポートを渡せる、省略時はデフォルトのポートを使う
  def initialize(left_motor_port=:port_c,
                 right_motor_port=:port_b,
                 color_port=:port_3)
    @left_wheel = EV3RT_TECS::Motor.new(left_motor_port)
    @right_wheel = EV3RT_TECS::Motor.new(right_motor_port)
    @color_sensor = EV3RT_TECS::ColorSensor.new(color_port)
  end

  def init
    Util::Display.init(self.class.to_s)
  end

  def terminate
    Util::Display.message("Stopped", 2)
    @right_wheel.stop
    @left_wheel.stop
  end
  
  def calc_prop_value
    kp = 0.83        # <1>
    target = 10      # <2>
    bias = 0
    diff = @color_sensor.reflect - target  # <3>
    result = kp * diff + bias             # <4>
  end

  def run
    Util::Display.message("running...", 2)
    turn = calc_prop_value
    pwm_l = PWM - turn
    pwm_r = PWM + turn
    @left_wheel.power = pwm_l
    @right_wheel.power = pwm_r
  end

end

File: Document/beginners/sample_java/sample04/src/jp/etrobo/ev3/sample04/app/Tracer.java
/*
 *  Tracer.java (for leJOS EV3)
 *  Created on: 2016/06/11
 *  Copyright (c) 2016 Embedded Technology Software Design Robot Contest
 */
package jp.etrobo.ev3.sample04.app;

import jp.etrobo.ev3.sample04.ColorSensor;
import lejos.hardware.lcd.LCD;
import lejos.hardware.port.MotorPort;
import lejos.hardware.port.SensorPort;
import lejos.hardware.port.TachoMotorPort;

public class Tracer {
    private TachoMotorPort leftWheel;
    private TachoMotorPort rightWheel;
    private ColorSensor    colorSensor;

	private static final int pwm = TachoMotorPort.MAX_POWER / 6;
	private static final float Kp = 83.0F;
	private static final float target = 0.1F;
	private static final float bias = 0.0F;

    public Tracer(){
    	leftWheel = MotorPort.C.open(TachoMotorPort.class);
    	rightWheel = MotorPort.B.open(TachoMotorPort.class);
    	colorSensor = new ColorSensor(SensorPort.S3);
    }

    public void init(){
    	LCD.drawString("Tracer", 0, 2);
    }
    public void run(){
    	LCD.drawString("running...", 0, 4);
    	float diff = colorSensor.getBrightness() - target;
    	float turn = Kp * diff + bias;
        leftWheel.controlMotor((int)(pwm - turn), TachoMotorPort.PWM_FLOAT);
        rightWheel.controlMotor((int)(pwm + turn), TachoMotorPort.PWM_FLOAT);
    }
    public void terminate(){
        LCD.drawString("Stopped.", 0, 4);
        leftWheel.setPWMMode(TachoMotorPort.STOP);
        rightWheel.setPWMMode(TachoMotorPort.STOP);

    }
}
